### ISSUE TITLE
Provide device-authentication workflow for GitHub?

### ISSUE THREAD

**Issue by cboettig** [r-lib/gitcreds#61]
Apologies if this is already supported here or elsewhere! Please let me know!  I think it would be great if `gitcreds` could provide the device authentication workflow for GitHub, such as we see in the Python package https://github.com/jupyterhub/gh-scoped-creds (and also built into MS's Code Server I believe), in which a user requests a short-lived (8hr) scoped credential authenticated with a OTP.  

This provides a much simpler workflow for users creating tokens than learning to do so manually in the GitHub web interface.  It also promotes the use of short-lived tokens, consistent with the best-practices we see in other systems, especially with the rise of hosted or cloud-based compute platforms.

you know I'm no security expert, but it seems that current practices in our R community are still built around more long-lived and widely scoped tokens, while elsewhere everything is migrating towards more short-lived and narrowly scoped ones.  (I recognize that this request is more GitHub specific than being generally about git, just wasn't sure where best to raise it).

**Comment by gaborcsardi** [comment]
Seems like gh-scoped-creds sets up a global git credential helper for github.com, so if you already use that, there is nothing more to do in R, gitcreds will automatically pick up and use those credentials.

**Comment by cboettig** [comment]
yes, that true in settings where students are working in python already anyway.  

but when I'm just teaching in R it would be nice to have a solution that doesn't require messing around with setting up a python environment first, right?  It's just [an API call and writing to a file](https://github.com/jupyterhub/gh-scoped-creds/blob/main/gh_scoped_creds/__init__.py)

**Comment by rstub** [comment]
A similar issue exists when using the otherwise extremely useful git credential helper from https://github.com/hickford/git-credential-oauth. From what I can tell, `gitcreds::gitcreds_get()` at some point calls `git credential fill` with `url=http://github.com` as input. The OAuth based credential helper than sends instructions to `STDERR` which are swallowed here. In the case of a desktop machine this is no big problem, since these instructions are of the form
```
Please complete authentication in your browser...
https://github.com/login/oauth/authorize?client_id=[...]
```
but one gets forwarded to that URL in the system browser anyway, so nothing is lost.

On a headless machine, e.g. using RStudio Server or Posit Workbench, the instructions are crucial, though:

```
Please enter code 897B-EA6C at https://github.com/login/device
```

If one does not follow these instructions, one cannot complete the sign-in and `gitcreds_get()` hangs (probably) indefinitely.

Workaround: Find the file `/tmp/Rtmp<random>/gitcreds-stderr-<random>` and follow the instructions in there.

**Comment by gaborcsardi** [comment]
@rstub Can you please open another issue for that, it seems like a different problem to me. Thanks.
